test/sol5/utils4.js

```diff
@@ -0,0 +1,320 @@
+        let tx1 = await utils4.mockCheckGetUpdateDecimals(token.address)
+        let tx2 = await utils4.mockCheckGetUpdateDecimals(token.address)
+        Helper.assertGreater(tx1.receipt.gasUsed, tx2.receipt.gasUsed)
+    });
+
+    it("should check get decimals for token without decimals API. see reverts", async function () {
+        let tokenNoDecimal = await TokenNoDecimal.new("noDec", "dec", 18);
+
+        //now get deicmals to see values
+        await expectRevert.unspecified(utils4.mockSetDecimals(tokenNoDecimal.address));
+        await expectRevert.unspecified(utils4.mockGetDecimals(tokenNoDecimal.address));
+        await expectRevert.unspecified(utils4.mockCheckGetUpdateDecimals(tokenNoDecimal.address));
+    });
+
+    it("check dest qty calculation for high quantities.", async function () {
+        let token1 = await TestToken.new("Token1", "1", 10)
+        let token2 = await TestToken.new("Token2", "2", 20)
+        let srcQty = MAX_QTY.div(new BN(2));
+        let rate = MAX_RATE;
+
+        //first check when dest decimals > src decimals
+        let srcDecimal = await token1.decimals();
+        let dstDecimal = await token2.decimals();
+        let srcAddress = token1.address;
+        let dstAddress = token2.address;
+
+        //should work with max Qty
+        let expectedDestQty = Helper.calcDstQty(srcQty, srcDecimal, dstDecimal, rate);
+        let reportedDstQty = await utils4.mockCalcDstQty(srcQty, srcDecimal, dstDecimal, rate);
+        Helper.assertEqual(expectedDestQty, reportedDstQty, "unexpected dst qty");
+        Helper.assertEqual(
+            expectedDestQty,
+            await utils4.mockCalcDestAmount(srcAddress, dstAddress, srcQty, rate),
+            "unexpected destamount"
+        );
+
+        //next check when dest decimals > src decimals
+        srcDecimal = await token2.decimals();
+        dstDecimal = await token1.decimals();
+        srcAddress = token2.address;
+        dstAddress = token1.address;
+
+        //should work with max Qty
+        expectedDestQty = Helper.calcDstQty(srcQty, srcDecimal, dstDecimal, rate);
+        reportedDstQty = await utils4.mockCalcDstQty(srcQty, srcDecimal, dstDecimal, rate);
+        Helper.assertEqual(expectedDestQty, reportedDstQty, "unexpected dst qty");
+        Helper.assertEqual(
+            expectedDestQty,
+            await utils4.mockCalcDestAmount(srcAddress, dstAddress, srcQty, rate),
+            "unexpected destamount"
+        );
+
+        //should revert
+        rate = MAX_RATE.add(new BN(1));
+        await expectRevert(
+            utils4.mockCalcDstQty(srcQty, srcDecimal, dstDecimal, rate),
+            "rate > MAX_RATE"
+        )
+        rate = MAX_RATE
+        srcQty = MAX_QTY.add(new BN(1))
+        await expectRevert(
+            utils4.mockCalcDstQty(srcQty, srcDecimal, dstDecimal, rate),
+            "srcQty > MAX_QTY"
+        )
+    });
+
+
+    it("check src qty calculation for high quantities.", async function () {
```

Comment:
please split into more tests.
per test need to test 1 thing.
so can split for srcDecimal > destDecimal.
destDecimals > srcDecimals
check revert

Reply:
break into smaller test